libs/langchain/langchain/memory/chat_message_histories/singlestoredb.py

```diff
@@ -0,0 +1,263 @@
+            results_type (str, optional): Determines the structure of the query results:
+                tuples, namedtuples, dicts.
+            results_format (str, optional): Deprecated. This option has been renamed to
+                results_type.
+
+        Examples:
+            Basic Usage:
+
+            .. code-block:: python
+
+                from langchain.memory.chat_message_histories import (
+                    SingleStoreDBChatMessageHistory
+                )
+
+                message_history = SingleStoreDBChatMessageHistory(
+                    session_id="my-session",
+                    host="https://user:password@127.0.0.1:3306/database"
+                )
+
+            Advanced Usage:
+
+            .. code-block:: python
+
+                from langchain.memory.chat_message_histories import (
+                    SingleStoreDBChatMessageHistory
+                )
+
+                message_history = SingleStoreDBChatMessageHistory(
+                    session_id="my-session",
+                    host="127.0.0.1",
+                    port=3306,
+                    user="user",
+                    password="password",
+                    database="db",
+                    table_name="my_custom_table",
+                    pool_size=10,
+                    timeout=60,
+                )
+
+            Using environment variables:
+
+            .. code-block:: python
+
+                from langchain.memory.chat_message_histories import (
+                    SingleStoreDBChatMessageHistory
+                )
+
+                os.environ['SINGLESTOREDB_URL'] = 'me:p455w0rd@s2-host.com/my_db'
+                message_history = SingleStoreDBChatMessageHistory("my-session")
+        """
+
+        self.table_name = table_name
+        self.session_id = session_id
+        self.id_field = id_field
+        self.session_id_field = session_id_field
+        self.message_field = message_field
+
+        """Pass the rest of the kwargs to the connection."""
+        self.connection_kwargs = kwargs
+
+        """Add connection attributes to the connection kwargs."""
+        if "conn_attrs" not in self.connection_kwargs:
+            self.connection_kwargs["conn_attrs"] = dict()
+
+        self.connection_kwargs["conn_attrs"]["_connector_name"] = "langchain python sdk"
+        self.connection_kwargs["conn_attrs"]["_connector_version"] = "1.0.1"
+
+        """Create a connection pool."""
+        try:
+            from sqlalchemy.pool import QueuePool
+        except ImportError:
+            raise ImportError(
+                "Could not import sqlalchemy.pool python package. "
+                "Please install it with `pip install singlestoredb`."
+            )
+
+        self.connection_pool = QueuePool(
+            self._get_connection,
+            max_overflow=max_overflow,
+            pool_size=pool_size,
+            timeout=timeout,
+        )
+        self._create_table_if_not_exists()
```

Comment:
Perhaps add an opt out option for the check as it involves an extra network request on instantiation that may not necessary. It's unclear to me what the life cycle of the object is in user applications. This code assumes that the user instantiates the object infrequently, which may or may not be the case

Reply:
I rewrote it in a `lazy` way: the create_table will be called with the first call that touches the data.